ping/ping_json.c

```diff
@@ -0,0 +1,217 @@
+/*
+ * SPDX-License-Identifier: BSD-3-Clause
+ * Copyright (c) 2024 Georg Pfuetzenreuter <mail+ip@georg-pfuetzenreuter.net>
+ */
+
+#include "ping.h"
+#include "stdarg.h"
+
+static inline void json_emergency(char *msg) {
+	error(1, 0, "%s", msg);
+}
+
+static inline void test_buflen(int have) {
+	if (have < 0 || have >= PING_JSON_MAX)
+		json_emergency("Overflow during JSON construction.");
+}
+
+static void json_end_array(char *part) {
+	size_t curlen = strlen(part);
+	test_buflen(snprintf(part + curlen, PING_JSON_MAX - curlen, "]"));
+}
+
+static void json_start_object(char *part) {
+	test_buflen(snprintf(part, PING_JSON_MAX, "{"));
+}
+
+static void json_end_object(char *part) {
+	size_t curlen = strlen(part);
+	test_buflen(snprintf(part + curlen, PING_JSON_MAX - curlen, "}"));
+}
+
+static void json_continue(char *part) {
+	size_t curlen = strlen(part);
+	test_buflen(snprintf(part + curlen, PING_JSON_MAX - curlen, ", "));
+}
+
+static void json_kv_str(char *part, char *key, char *value) {
+	size_t curlen = strlen(part);
+	test_buflen(snprintf(part + curlen, PING_JSON_MAX - curlen, "\"%s\": \"%s\"", key, value));
+}
+
+static void json_kv_str_continue(char *part, char *key, char *value) {
+	json_kv_str(part, key, value);
+	json_continue(part);
+}
+
+static void json_kv_int(char *part, char *key, int value) {
+	size_t curlen = strlen(part);
+	test_buflen(snprintf(part + curlen, PING_JSON_MAX - curlen, "\"%s\": %d", key, value));
+}
+
+static void json_kv_uint(char *part, char *key, unsigned int value) {
+	size_t curlen = strlen(part);
+	test_buflen(snprintf(part + curlen, PING_JSON_MAX - curlen, "\"%s\": %u", key, value));
+}
+
+static void json_kv_int_continue(char *part, char *key, int value) {
+	json_kv_int(part, key, value);
+	json_continue(part);
+}
+
+/* so far only supports arrays of strings */
+static void json_kv_array(char *part, char *key, va_list ap) {
+	size_t curlen = strlen(part);
+	test_buflen(snprintf(part + curlen, PING_JSON_MAX - curlen, "\"%s\": [", key));
+
+	char * value;
```

Comment:
A bit more consistent formatting would be nice. `*` should be without following space `char *value`. I would also name it just `val` (clear enough).

Reply:
My bad, aligned now, and s/value/val/g.